File: src/main/java/org/gecko/actions/CreateSystemConnectionViewModelElementAction.java
package org.gecko.actions;

import org.gecko.exceptions.GeckoException;
import org.gecko.viewmodel.GeckoViewModel;
import org.gecko.viewmodel.PortViewModel;
import org.gecko.viewmodel.SystemConnectionViewModel;
import org.gecko.viewmodel.SystemViewModel;

public class CreateSystemConnectionViewModelElementAction extends Action {

    private final GeckoViewModel geckoViewModel;
    private final PortViewModel source;
    private final PortViewModel destination;
    private SystemConnectionViewModel createdSystemConnectionViewModel;

    CreateSystemConnectionViewModelElementAction(
        GeckoViewModel geckoViewModel, PortViewModel source, PortViewModel destination) {
        this.geckoViewModel = geckoViewModel;
        this.source = source;
        this.destination = destination;
    }

    @Override
    boolean run() throws GeckoException {
        SystemViewModel currentParentSystem = geckoViewModel.getCurrentEditor().getCurrentSystem();
        SystemViewModel sourceSystem = (SystemViewModel) geckoViewModel.getViewModelElement(
            currentParentSystem.getTarget().getChildSystemWithVariable(source.getTarget()));
        SystemViewModel destinationSystem = (SystemViewModel) geckoViewModel.getViewModelElement(
            currentParentSystem.getTarget().getChildSystemWithVariable(destination.getTarget()));
        if (sourceSystem == null || destinationSystem == null || sourceSystem == destinationSystem) {
            return false;
        }

        if (destination.getTarget().isHasIncomingConnection()) {
            return false;
        }

        createdSystemConnectionViewModel = geckoViewModel.getViewModelFactory()
            .createSystemConnectionViewModelIn(currentParentSystem, source, destination);
        createdSystemConnectionViewModel.setSource(source);
        createdSystemConnectionViewModel.setDestination(destination);
        createdSystemConnectionViewModel.updateTarget();
        ActionManager actionManager = geckoViewModel.getActionManager();
        actionManager.run(actionManager.getActionFactory().createSelectAction(createdSystemConnectionViewModel, true));
        return true;
    }

    @Override
    Action getUndoAction(ActionFactory actionFactory) {
        return actionFactory.createDeletePositionableViewModelElementAction(createdSystemConnectionViewModel);
    }
}

File: src/main/kotlin/org/gecko/actions/CreateSystemConnectionViewModelElementAction.kt
package org.gecko.actions

import org.gecko.exceptions.GeckoException
import org.gecko.viewmodel.*

/**
 * A concrete representation of an [Action] that creates an [SystemConnectionViewModel] in the
 * current-[System] with given source- and destination-[Port]s through the
 * [ViewModelFactory][org.gecko.viewmodel.ViewModelFactory] of the [GModel].
 */
class CreateSystemConnectionViewModelElementAction internal constructor(
    val gModel: GModel,
    val source: Port,
    val destination: Port
) : Action() {
    var createdSystemConnectionViewModel: SystemConnectionViewModel? = null

    @Throws(GeckoException::class)
    override fun run(): Boolean {
        val currentParentSystem = gModel.currentEditor!!.currentSystem
        if (!isConnectingAllowed(
                source,
                destination,
                gModel.getSystemViewModelWithPort(source),
                gModel.getSystemViewModelWithPort(destination),
                currentParentSystem,
                null
            )
        ) {
            return false
        }

        createdSystemConnectionViewModel = gModel.viewModelFactory
            .createSystemConnectionViewModelIn(currentParentSystem, source, destination)

        val actionManager = gModel.actionManager
        actionManager.run(actionManager.actionFactory.createSelectAction(createdSystemConnectionViewModel!!, true))
        return true
    }

    override fun getUndoAction(actionFactory: ActionFactory): Action {
        return actionFactory.createDeleteAction(createdSystemConnectionViewModel!!)
    }
}
